qbraid/transpiler/conversions/openqasm3/openqasm3_to_cudaq.py

```diff
@@ -0,0 +1,172 @@
+# or later. You can redistribute and/or modify it under the terms of the GPL v3.
+# See the LICENSE file in the project root or <https://www.gnu.org/licenses/gpl-3.0.html>.
+#
+# THERE IS NO WARRANTY for the qBraid-SDK, as per Section 15 of the GPL v3.
+
+"""
+Module containing OpenQASM to CUDA-Q conversion function
+
+"""
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Union
+
+import cudaq
+from openqasm3 import ast
+import pyqasm
+
+from qbraid.transpiler.annotations import weight
+from qbraid.transpiler.exceptions import ProgramConversionError
+
+if TYPE_CHECKING:
+    from cudaq import PyKernel, QuakeValue
+
+
+def gate_kernel(name: str, *args) -> PyKernel:
+    if name in ["x", "y", "z", "h", "s", "t"]:
+        size, argc = 1, 0
+    elif name in ["rx", "ry", "rz"]:
+        size, argc = 1, 1
+    elif name in ["swap"]:
+        size, argc = 2, 0
+    elif name in ["u3"]:
+        size, argc = 3, 3
+    else:
+        raise ProgramConversionError(f"Unsupported gate: {name}")
+
+    kernel, *qrefs = cudaq.make_kernel(*[cudaq.qubit for _ in range(size)])
+
+    if len(args) != argc:
+        raise ProgramConversionError(
+            f"Gate {name} requires {argc} args but only {len(args)} were provided: {args}"
+        )
+
+    op = getattr(kernel, name)
+
+    if len(args) > 0:
+        op(*args, *qrefs)
+    else:
+        op(*qrefs)
+
+    return kernel
+
+
+def identifier_lookup(cbit: Union[ast.IndexedIdentifier, ast.Identifier]) -> str:
+    """Unpacks the identifier name"""
+    if isinstance(cbit, ast.IndexedIdentifier):
+        return cbit.name.name
+
+    return cbit.name
+
+
+@weight(1)
+def openqasm3_to_cudaq(program: ast.Program) -> PyKernel:
+    """Returns a CUDA-Q kernel representing the input OpenQASM program.
+
+    Args:
+        qasm (str or ast.Program): OpenQASM program to convert to CUDA-Q kernel.
+
+    Returns:
+        kernel: CUDA-Q kernel equivalent to input OpenQASM string.
+    """
+    try:
+        pyqasm.load(program).validate()
+    except Exception as e:
+        raise ProgramConversionError("PyQasm program is not well-formed") from e
+        
+    
+    kernel = cudaq.make_kernel()
+    ctx = {}
+
+    def qubit_lookup(qubit: Union[ast.IndexedIdentifier, ast.Identifier]) -> QuakeValue:
+        if isinstance(qubit, ast.IndexedIdentifier):
+            if len(qubit.indices) > 1:
+                raise ProgramConversionError(
+                    f"Multi-dimensional array indexing is unsupported: {qubit.indices}"
+                )
+
+            inds = qubit.indices[0]
+            if isinstance(inds, ast.DiscreteSet):
+                inds = inds.values
+
+            if len(inds) > 1:
+                raise ProgramConversionError(
+                    f"Discrete set or multi-integer indexing is unsupported: {inds}"
+                )
+
+            ind = inds[0]
+            if not isinstance(ind, ast.IntegerLiteral):
```

Comment:
We don't need this validation, `pyqasm` handles the indexing for all indexed identifiers

Reply:
Got it, will remove